addons/website_sale/tests/test_sale_process.py

```diff
@@ -274,6 +275,43 @@ def test_03_public_user_address_and_company(self):
             self.WebsiteSaleController.address(**self.default_address_values)
             self.assertEqual(new_partner.company_id, self.website.company_id, "Public user edited billing (the partner itself) should not get its company modified.")
 
+    def test_03_carrier_rate_on_shipping_address_change(self):
+        ''' Test that when a shipping address is changed the price of delivery is recalculated
+        and updated on the order.'''
+        # Create a dummy carrier.
+        self.product_delivery_dummy = self.env['product.product'].create(
+            {
+                'name': 'Dummy',
+                'type': 'service',
+                'categ_id': self.env.ref('delivery.product_category_deliveries').id,
+            }
+        )
+        carrier = self.env['delivery.carrier'].create(
+            {
+                'name': 'Fixed',
+                'product_id': self.product_delivery_dummy.id,
+                'website_published': True,
+            }
+        )
+        p = self.env.user.partner_id
+        order = self._create_so(p.id)
+        order.carrier_id = carrier.id  # Set the carrier on the order.
+        shipping_partner_values = {'name': 'dummy', 'parent_id': p.id, 'type': 'delivery'}
+        shipping_partner = self.env["res.partner"].create(shipping_partner_values)
+        with MockRequest(self.env, website=self.website, sale_order_id=order.id):
+            with patch(
+                'odoo.addons.delivery.models.delivery_carrier.DeliveryCarrier.rate_shipment'
+            ) as rate:
+                rate.return_value = {'success': True, 'price': 10, 'warning_message': ''}
+                # Change a shipping address of the order in the checkout.
+                self.WebsiteSaleController.checkout_values(partner_id=shipping_partner.id)
+                msg = "The carrier rate must be recalculated when shipping address is changed."
+                self.assertTrue(rate.called, msg=msg)
```

Comment:
Yes, but if for some reasons in the future this function will be called more than once and it will break the test. And it is not clear why it should be called only once, no?

Reply:
In that case we could update the test, but you could use `assertGreaterEqual` if you want.

Reply:
yes, thank you!